### PR TITLE
Making `make install` work better by default.

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

Making `make install` a much better sane default to start local dev environments.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: .github/workflows/tests.yaml [hunk 1]
```diff
@@ -68,7 +68,7 @@ jobs:
             ~/.cargo/git
       - name: Install
         run: |
-          make install
+          make install-cpu
```
Comment (Narsil) [inline .github/workflows/tests.yaml:71]: Simple server tests are running on non accelerated hardwared, therefore cannot install the kernels.

File: Cargo.toml [hunk 2]
```diff
@@ -19,6 +19,10 @@ tokenizers = { version = "0.19.1", features = ["http"] }
 hf-hub = { version = "0.3.1", features = ["tokio"] }
 
 [profile.release]
+incremental = true
```
Comment (Narsil) [inline Cargo.toml:22]: This speeds up devs by improving significantly build times.

File: Dockerfile [hunk 3]
```diff
@@ -193,7 +193,7 @@ COPY --from=flash-att-builder /usr/src/flash-attention/csrc/layer_norm/build/lib
 COPY --from=flash-att-builder /usr/src/flash-attention/csrc/rotary/build/lib.linux-x86_64-cpython-310 /opt/conda/lib/python3.10/site-packages
 
 # Copy build artifacts from flash attention v2 builder
-COPY --from=flash-att-v2-builder /usr/src/flash-attention-v2/build/lib.linux-x86_64-cpython-310 /opt/conda/lib/python3.10/site-packages
+COPY --from=flash-att-v2-builder /opt/conda/lib/python3.10/site-packages/flash_attn_2_cuda.cpython-310-x86_64-linux-gnu.so /opt/conda/lib/python3.10/site-packages
```
Comment (Narsil) [inline Dockerfile:196]: `pip installing` means the library is in this location instead.

File: Makefile [hunk 4]
```diff
@@ -1,12 +1,8 @@
...
 	cd server && make install
 
-install-custom-kernels:
-	if [ "$$BUILD_EXTENSIONS" = "True" ]; then cd server/custom_kernels && python setup.py install; else echo "Custom kernels are disabled, you need to set the BUILD_EXTENSIONS environment variable to 'True' in order to build them. (Please read the docs, kernels might not work on all hardware)"; fi
-
-install-integration-tests:
-	cd integration-tests && pip install -r requirements.txt
-	cd clients/python && pip install .
+install-server-cpu:
+	cd server && make install-server
```
Comment (Narsil) [inline Makefile:5]: BUILD_EXTENSIONS is removed (it's not used anymore anyway, only for bloom and it's not even guaranteed to work on all hardware)

File: server/Makefile [hunk 5]
```diff
@@ -10,18 +10,26 @@ unit-tests:
 
 gen-server:
 	# Compile protos
-	pip install grpcio-tools==1.51.1 mypy-protobuf==3.4.0 'types-protobuf>=3.20.4' --no-cache-dir
+	pip install grpcio-tools==1.62.2 mypy-protobuf==3.6.0 'types-protobuf' --no-cache-dir
```
Comment (Narsil) [inline server/Makefile:13]: In line with `pyproject.toml`

File: server/Makefile-flash-att [hunk 6]
```diff
@@ -1,16 +1,14 @@
...
-	pip install -U packaging ninja  --no-cache-dir
-	git clone https://github.com/HazyResearch/flash-attention.git
-
-build-flash-attention: flash-attention
-	cd flash-attention && git fetch && git checkout $(flash_att_commit)
-	cd flash-attention && python setup.py build
-	cd flash-attention/csrc/rotary && python setup.py build
-	cd flash-attention/csrc/layer_norm && python setup.py build
+build-flash-attention:
+	if [ ! -d 'flash-attention' ]; then \
```
Comment (Narsil) [inline server/Makefile-flash-att:4]: Very dumb protection for recompiling all the time, at least it doesn't fail when you rerun the command.

Comment (Narsil) [reply server/Makefile-flash-att:4]: Keeing `build-` variants only because of the dockerfiles.

File: server/Makefile-flash-att-v2 [hunk 7]
```diff
@@ -1,29 +1,18 @@
-flash_att_v2_commit_cuda := v2.5.8
+flash_att_v2_commit_cuda := v2.5.9.post1
 flash_att_v2_commit_rocm := 2554f490101742ccdc56620a938f847f61754be6
 
+build-flash-attention-v2-cuda:
+	pip install -U packaging wheel
```
Comment (Narsil) [inline server/Makefile-flash-att-v2:5]: Wheel is necessary and `flash-attn` doesn't respect pep517 (https://github.com/Dao-AILab/flash-attention/issues/453) so nothing really better is doable I feel

**Review by drbh — APPROVED** [review]
looks good to me! great change!

**Pull request merged**